## **CLASS 22: PANDAS CONTD: 30TH JULY 2022**

DATASET USED:

https://drive.google.com/file/d/1FRnbQqgK6S-mr9StNF5MurSaXhah9uoJ/view?usp=sharing


https://drive.google.com/file/d/11rPJXr-9VoeV2Pi0RSfVehw1tbQOAjLI/view?usp=sharing

In [2]:
import pandas as pd

In [3]:
df = pd.read_csv("/content/sample_data/sales_data_final.csv")

In [4]:
df.head(3) #Top 3 Rows

order_id order_date ship_date       ship_mode    customer_name  \
0   AG-2011-2040   1/1/2011  1/6/2011  Standard Class  Toby Braunhardt   
1  IN-2011-47883   1/1/2011  1/8/2011  Standard Class      Joseph Holt   
2   HU-2011-1220   1/1/2011  1/5/2011    Second Class    Annie Thurman   

    segment            state    country  market   region  ...  \
0  Consumer      Constantine    Algeria  Africa   Africa  ...   
1  Consumer  New South Wales  Australia    APAC  Oceania  ...   
2  Consumer         Budapest    Hungary    EMEA     EMEA  ...   

          category sub_category              product_name sales quantity  \
0  Office Supplies      Storage       Tenex Lockers, Blue   408        2   
1  Office Supplies     Supplies  Acme Trimmer, High Speed   120        3   
2  Office Supplies      Storage   Tenex Box, Single Width    66        4   

   discount   profit  shipping_cost  order_priority  year  
0       0.0  106.140          35.46          Medium  2011  
1       0.1   36.036           9.72          Medium  2011  
2       0.0   29.640           8.17            High  2011  

[3 rows x 21 columns]

In [5]:
df.tail(3) #Last 3 Rows

order_id  order_date ship_date       ship_mode     customer_name  \
51287  MX-2014-110527  12/31/2014  1/2/2015    Second Class  Charlotte Melton   
51288  MX-2014-114783  12/31/2014  1/6/2015  Standard Class     Tamara Dahlen   
51289  CA-2014-156720  12/31/2014  1/4/2015  Standard Class     Jill Matthias   

        segment      state        country market   region  ...  \
51287  Consumer    Managua      Nicaragua  LATAM  Central  ...   
51288  Consumer  Chihuahua         Mexico  LATAM    North  ...   
51289  Consumer   Colorado  United States     US     West  ...   

              category sub_category                            product_name  \
51287  Office Supplies       Labels  Hon Color Coded Labels, 5000 Label Set   
51288  Office Supplies       Labels  Hon Legal Exhibit Labels, Alphabetical   
51289  Office Supplies    Fasteners                     Bagged Rubber Bands   

      sales quantity  discount   profit  shipping_cost  order_priority  year  
51287    26        3       0.0  12.3600           0.35          Medium  2014  
51288     7        1       0.0   0.5600           0.20          Medium  2014  
51289     3        3       0.2  -0.6048           0.17          Medium  2014  

[3 rows x 21 columns]

In [6]:
df.columns #List of all columns

Index(['order_id', 'order_date', 'ship_date', 'ship_mode', 'customer_name',
       'segment', 'state', 'country', 'market', 'region', 'product_id',
       'category', 'sub_category', 'product_name', 'sales', 'quantity',
       'discount', 'profit', 'shipping_cost', 'order_priority', 'year'],
      dtype='object')

In [7]:
df['order_date'] #Extracting Single Column

0          1/1/2011
1          1/1/2011
2          1/1/2011
3          1/1/2011
4          1/1/2011
            ...    
51285    12/31/2014
51286    12/31/2014
51287    12/31/2014
51288    12/31/2014
51289    12/31/2014
Name: order_date, Length: 51290, dtype: object

In [8]:
type(df['order_date']) #Series (Equivalent to list). Pandas has only 2 data types: Series and DataFrame

pandas.core.series.Series

In [9]:
df.dtypes #Data Types

order_id           object
order_date         object
ship_date          object
ship_mode          object
customer_name      object
segment            object
state              object
country            object
market             object
region             object
product_id         object
category           object
sub_category       object
product_name       object
sales              object
quantity            int64
discount          float64
profit            float64
shipping_cost     float64
order_priority     object
year                int64
dtype: object

In [10]:
df['profit']

0        106.1400
1         36.0360
2         29.6400
3        -26.0550
4         37.7700
           ...   
51285      4.5188
51286      0.4200
51287     12.3600
51288      0.5600
51289     -0.6048
Name: profit, Length: 51290, dtype: float64

In [11]:
df[['profit','year']] #Fetching Multiple Columns

profit  year
0      106.1400  2011
1       36.0360  2011
2       29.6400  2011
3      -26.0550  2011
4       37.7700  2011
...         ...   ...
51285    4.5188  2014
51286    0.4200  2014
51287   12.3600  2014
51288    0.5600  2014
51289   -0.6048  2014

[51290 rows x 2 columns]

In [12]:
type([[df['profit']]]) #For multiple columns fetching, we need to pass as a list otherwise we will get an error

list

In [13]:
type(df['profit'])

pandas.core.series.Series

In [14]:
df.columns

Index(['order_id', 'order_date', 'ship_date', 'ship_mode', 'customer_name',
       'segment', 'state', 'country', 'market', 'region', 'product_id',
       'category', 'sub_category', 'product_name', 'sales', 'quantity',
       'discount', 'profit', 'shipping_cost', 'order_priority', 'year'],
      dtype='object')

In [15]:
df[['order_id','product_id','quantity']]

order_id        product_id  quantity
0         AG-2011-2040  OFF-TEN-10000025         2
1        IN-2011-47883   OFF-SU-10000618         3
2         HU-2011-1220  OFF-TEN-10001585         4
3      IT-2011-3647632   OFF-PA-10001492         3
4        IN-2011-47883   FUR-FU-10003447         5
...                ...               ...       ...
51285   CA-2014-115427   OFF-BI-10002103         2
51286     MO-2014-2560  OFF-WIL-10001069         1
51287   MX-2014-110527   OFF-LA-10004182         3
51288   MX-2014-114783   OFF-LA-10000413         1
51289   CA-2014-156720   OFF-FA-10003472         3

[51290 rows x 3 columns]

In [16]:
df.describe() #Descriptive Statistics of Data: Only about integer and Float data types, description will be mentioned
#Count here shows that there is no Null Values for the columns mentioned
#Pandas Profiling will give us detailed analytics on the data by using a single line of code
#Standard Deviation: Deviation from Mean: Root of Variance
#25th Percentile Data: Data is arranged in ascending order and then 25th percent of data

quantity      discount        profit  shipping_cost          year
count  51290.000000  51290.000000  51290.000000   51290.000000  51290.000000
mean       3.476545      0.142908     28.641740      26.375915   2012.777208
std        2.278766      0.212280    174.424113      57.296804      1.098931
min        1.000000      0.000000  -6599.978000       0.000000   2011.000000
25%        2.000000      0.000000      0.000000       2.610000   2012.000000
50%        3.000000      0.000000      9.240000       7.790000   2013.000000
75%        5.000000      0.200000     36.810000      24.450000   2014.000000
max       14.000000      0.850000   8399.976000     933.570000   2014.000000

In [17]:
df.dtypes == 'object' 

order_id           True
order_date         True
ship_date          True
ship_mode          True
customer_name      True
segment            True
state              True
country            True
market             True
region             True
product_id         True
category           True
sub_category       True
product_name       True
sales              True
quantity          False
discount          False
profit            False
shipping_cost     False
order_priority     True
year              False
dtype: bool

In [18]:
type(df.dtypes) #series (Like List)

pandas.core.series.Series

In [19]:
df.dtypes[df.dtypes == 'object'] #Only Object data type shown

order_id          object
order_date        object
ship_date         object
ship_mode         object
customer_name     object
segment           object
state             object
country           object
market            object
region            object
product_id        object
category          object
sub_category      object
product_name      object
sales             object
order_priority    object
dtype: object

In [20]:
df.dtypes[df.dtypes == 'object'].index #List of all the columns having data type object. This statement outputs a list

Index(['order_id', 'order_date', 'ship_date', 'ship_mode', 'customer_name',
       'segment', 'state', 'country', 'market', 'region', 'product_id',
       'category', 'sub_category', 'product_name', 'sales', 'order_priority'],
      dtype='object')

In [21]:
df[df.dtypes[df.dtypes == 'object'].index] #Passing List int the dataframe gives us the data with specific columns

order_id  order_date ship_date       ship_mode  \
0         AG-2011-2040    1/1/2011  1/6/2011  Standard Class   
1        IN-2011-47883    1/1/2011  1/8/2011  Standard Class   
2         HU-2011-1220    1/1/2011  1/5/2011    Second Class   
3      IT-2011-3647632    1/1/2011  1/5/2011    Second Class   
4        IN-2011-47883    1/1/2011  1/8/2011  Standard Class   
...                ...         ...       ...             ...   
51285   CA-2014-115427  12/31/2014  1/4/2015  Standard Class   
51286     MO-2014-2560  12/31/2014  1/5/2015  Standard Class   
51287   MX-2014-110527  12/31/2014  1/2/2015    Second Class   
51288   MX-2014-114783  12/31/2014  1/6/2015  Standard Class   
51289   CA-2014-156720  12/31/2014  1/4/2015  Standard Class   

          customer_name      segment             state        country  market  \
0       Toby Braunhardt     Consumer       Constantine        Algeria  Africa   
1           Joseph Holt     Consumer   New South Wales      Australia    APAC   
2         Annie Thurman     Consumer          Budapest        Hungary    EMEA   
3          Eugene Moren  Home Office         Stockholm         Sweden      EU   
4           Joseph Holt     Consumer   New South Wales      Australia    APAC   
...                 ...          ...               ...            ...     ...   
51285        Erica Bern    Corporate        California  United States      US   
51286         Liz Preis     Consumer  Souss-Massa-Draƒ        Morocco  Africa   
51287  Charlotte Melton     Consumer           Managua      Nicaragua   LATAM   
51288     Tamara Dahlen     Consumer         Chihuahua         Mexico   LATAM   
51289     Jill Matthias     Consumer          Colorado  United States      US   

        region        product_id         category sub_category  \
0       Africa  OFF-TEN-10000025  Office Supplies      Storage   
1      Oceania   OFF-SU-10000618  Office Supplies     Supplies   
2         EMEA  OFF-TEN-10001585  Office Supplies      Storage   
3        North   OFF-PA-10001492  Office Supplies        Paper   
4      Oceania   FUR-FU-10003447        Furniture  Furnishings   
...        ...               ...              ...          ...   
51285     West   OFF-BI-10002103  Office Supplies      Binders   
51286   Africa  OFF-WIL-10001069  Office Supplies      Binders   
51287  Central   OFF-LA-10004182  Office Supplies       Labels   
51288    North   OFF-LA-10000413  Office Supplies       Labels   
51289     West   OFF-FA-10003472  Office Supplies    Fasteners   

                                          product_name sales order_priority  
0                                  Tenex Lockers, Blue   408         Medium  
1                             Acme Trimmer, High Speed   120         Medium  
2                              Tenex Box, Single Width    66           High  
3                          Enermax Note Cards, Premium    45           High  
4                           Eldon Light Bulb, Duo Pack   114         Medium  
...                                                ...   ...            ...  
51285  Cardinal Slant-D Ring Binder, Heavy Gauge Vinyl    14         Medium  
51286          Wilson Jones Hole Reinforcements, Clear     4         Medium  
51287           Hon Color Coded Labels, 5000 Label Set    26         Medium  
51288           Hon Legal Exhibit Labels, Alphabetical     7         Medium  
51289                              Bagged Rubber Bands     3         Medium  

[51290 rows x 16 columns]

In [22]:
df[df.dtypes[df.dtypes == 'object'].index].describe() #Description of Object Data type is possible

order_id order_date   ship_date       ship_mode  \
count            51290      51290       51290           51290   
unique           25035       1430        1464               4   
top     CA-2014-100111  6/18/2014  11/22/2014  Standard Class   
freq                14        135         130           30775   

          customer_name   segment       state        country market   region  \
count             51290     51290       51290          51290  51290    51290   
unique              795         3        1094            147      7       13   
top     Muhammed Yedwab  Consumer  California  United States   APAC  Central   
freq                108     26518        2001           9994  11002    11117   

             product_id         category sub_category product_name  sales  \
count             51290            51290        51290        51290  51290   
unique            10292                3           17         3788   2246   
top     OFF-AR-10003651  Office Supplies      Binders      Staples     13   
freq                 35            31273         6152          227    589   

       order_priority  
count           51290  
unique              4  
top            Medium  
freq            29433

In [23]:
df[df.dtypes[df.dtypes == 'float'].index] #Fetching Data for Float Datatype

discount    profit  shipping_cost
0           0.0  106.1400          35.46
1           0.1   36.0360           9.72
2           0.0   29.6400           8.17
3           0.5  -26.0550           4.82
4           0.1   37.7700           4.70
...         ...       ...            ...
51285       0.2    4.5188           0.89
51286       0.0    0.4200           0.49
51287       0.0   12.3600           0.35
51288       0.0    0.5600           0.20
51289       0.2   -0.6048           0.17

[51290 rows x 3 columns]

In [24]:
df.order_id

0           AG-2011-2040
1          IN-2011-47883
2           HU-2011-1220
3        IT-2011-3647632
4          IN-2011-47883
              ...       
51285     CA-2014-115427
51286       MO-2014-2560
51287     MX-2014-110527
51288     MX-2014-114783
51289     CA-2014-156720
Name: order_id, Length: 51290, dtype: object

In [25]:
df['order_id']

0           AG-2011-2040
1          IN-2011-47883
2           HU-2011-1220
3        IT-2011-3647632
4          IN-2011-47883
              ...       
51285     CA-2014-115427
51286       MO-2014-2560
51287     MX-2014-110527
51288     MX-2014-114783
51289     CA-2014-156720
Name: order_id, Length: 51290, dtype: object

In [26]:
df['order_id'][1:4] #List Operations can be done as Series is equivalent to list

1      IN-2011-47883
2       HU-2011-1220
3    IT-2011-3647632
Name: order_id, dtype: object

In [27]:
df['order_id'][1:40:2] #With jump Size

1       IN-2011-47883
3     IT-2011-3647632
5       IN-2011-47883
7       IN-2011-79397
9        IZ-2011-4680
11      IN-2011-65159
13      IN-2011-33652
15     MX-2011-160234
17      ID-2011-80230
19      ID-2011-12596
21       IR-2011-7690
23       TZ-2011-7370
25      IN-2011-65159
27     MX-2011-111255
29     MX-2011-158771
31     MX-2011-111255
33     US-2011-136007
35     MX-2011-159373
37    IT-2011-2942451
39      IN-2011-33036
Name: order_id, dtype: object

In [28]:
df['category']

0        Office Supplies
1        Office Supplies
2        Office Supplies
3        Office Supplies
4              Furniture
              ...       
51285    Office Supplies
51286    Office Supplies
51287    Office Supplies
51288    Office Supplies
51289    Office Supplies
Name: category, Length: 51290, dtype: object

In [29]:
df['category1'] = 'Gaurav'
df

order_id  order_date ship_date       ship_mode  \
0         AG-2011-2040    1/1/2011  1/6/2011  Standard Class   
1        IN-2011-47883    1/1/2011  1/8/2011  Standard Class   
2         HU-2011-1220    1/1/2011  1/5/2011    Second Class   
3      IT-2011-3647632    1/1/2011  1/5/2011    Second Class   
4        IN-2011-47883    1/1/2011  1/8/2011  Standard Class   
...                ...         ...       ...             ...   
51285   CA-2014-115427  12/31/2014  1/4/2015  Standard Class   
51286     MO-2014-2560  12/31/2014  1/5/2015  Standard Class   
51287   MX-2014-110527  12/31/2014  1/2/2015    Second Class   
51288   MX-2014-114783  12/31/2014  1/6/2015  Standard Class   
51289   CA-2014-156720  12/31/2014  1/4/2015  Standard Class   

          customer_name      segment             state        country  market  \
0       Toby Braunhardt     Consumer       Constantine        Algeria  Africa   
1           Joseph Holt     Consumer   New South Wales      Australia    APAC   
2         Annie Thurman     Consumer          Budapest        Hungary    EMEA   
3          Eugene Moren  Home Office         Stockholm         Sweden      EU   
4           Joseph Holt     Consumer   New South Wales      Australia    APAC   
...                 ...          ...               ...            ...     ...   
51285        Erica Bern    Corporate        California  United States      US   
51286         Liz Preis     Consumer  Souss-Massa-Draƒ        Morocco  Africa   
51287  Charlotte Melton     Consumer           Managua      Nicaragua   LATAM   
51288     Tamara Dahlen     Consumer         Chihuahua         Mexico   LATAM   
51289     Jill Matthias     Consumer          Colorado  United States      US   

        region  ... sub_category  \
0       Africa  ...      Storage   
1      Oceania  ...     Supplies   
2         EMEA  ...      Storage   
3        North  ...        Paper   
4      Oceania  ...  Furnishings   
...        ...  ...          ...   
51285     West  ...      Binders   
51286   Africa  ...      Binders   
51287  Central  ...       Labels   
51288    North  ...       Labels   
51289     West  ...    Fasteners   

                                          product_name sales quantity  \
0                                  Tenex Lockers, Blue   408        2   
1                             Acme Trimmer, High Speed   120        3   
2                              Tenex Box, Single Width    66        4   
3                          Enermax Note Cards, Premium    45        3   
4                           Eldon Light Bulb, Duo Pack   114        5   
...                                                ...   ...      ...   
51285  Cardinal Slant-D Ring Binder, Heavy Gauge Vinyl    14        2   
51286          Wilson Jones Hole Reinforcements, Clear     4        1   
51287           Hon Color Coded Labels, 5000 Label Set    26        3   
51288           Hon Legal Exhibit Labels, Alphabetical     7        1   
51289                              Bagged Rubber Bands     3        3   

      discount    profit  shipping_cost  order_priority  year category1  
0          0.0  106.1400          35.46          Medium  2011    Gaurav  
1          0.1   36.0360           9.72          Medium  2011    Gaurav  
2          0.0   29.6400           8.17            High  2011    Gaurav  
3          0.5  -26.0550           4.82            High  2011    Gaurav  
4          0.1   37.7700           4.70          Medium  2011    Gaurav  
...        ...       ...            ...             ...   ...       ...  
51285      0.2    4.5188           0.89          Medium  2014    Gaurav  
51286      0.0    0.4200           0.49          Medium  2014    Gaurav  
51287      0.0   12.3600           0.35          Medium  2014    Gaurav  
51288      0.0    0.5600           0.20          Medium  2014    Gaurav  
51289      0.2   -0.6048           0.17          Medium  2014    Gaurav  

[51290 rows x 22 columns]

In [30]:
df.columns

Index(['order_id', 'order_date', 'ship_date', 'ship_mode', 'customer_name',
       'segment', 'state', 'country', 'market', 'region', 'product_id',
       'category', 'sub_category', 'product_name', 'sales', 'quantity',
       'discount', 'profit', 'shipping_cost', 'order_priority', 'year',
       'category1'],
      dtype='object')

In [31]:
df['order_id'].isnull() #Data null or not

0        False
1        False
2        False
3        False
4        False
         ...  
51285    False
51286    False
51287    False
51288    False
51289    False
Name: order_id, Length: 51290, dtype: bool

In [32]:
df['order_id'].isnull().sum() #No Null Data present in the column

0

In [33]:
df1 = pd.read_csv("/content/sample_data/DressSales.csv")

In [34]:
df1.isnull()

Dress_ID  29/8/2013  31/8/2013  2/9/2013  4/9/2013  6/9/2013  8/9/2013  \
0       False      False      False     False     False     False     False   
1       False      False      False     False     False     False     False   
2       False      False      False     False     False     False     False   
3       False      False      False     False     False     False     False   
4       False      False      False     False     False     False     False   
..        ...        ...        ...       ...       ...       ...       ...   
495     False      False      False     False     False     False     False   
496     False      False      False     False     False     False     False   
497     False      False      False     False     False     False     False   
498     False      False      False     False     False     False     False   
499     False      False      False     False     False     False     False   

     10/9/2013  12/9/2013  14/9/2013  ...  Unnamed: 26  Unnamed: 27  \
0        False      False      False  ...         True         True   
1        False      False      False  ...         True         True   
2        False      False      False  ...         True         True   
3        False      False      False  ...         True         True   
4        False      False      False  ...         True         True   
..         ...        ...        ...  ...          ...          ...   
495      False      False      False  ...         True         True   
496      False      False      False  ...         True         True   
497      False      False      False  ...         True         True   
498      False      False      False  ...         True         True   
499      False      False      False  ...         True         True   

     Unnamed: 28  Unnamed: 29  Unnamed: 30  Unnamed: 31  Unnamed: 32  \
0           True         True         True         True         True   
1           True         True         True         True         True   
2           True         True         True         True         True   
3           True         True         True         True         True   
4           True         True         True         True         True   
..           ...          ...          ...          ...          ...   
495         True         True         True         True         True   
496         True         True         True         True         True   
497         True         True         True         True         True   
498         True         True         True         True         True   
499         True         True         True         True         True   

     Unnamed: 33  Unnamed: 34  Unnamed: 35  
0           True         True         True  
1           True         True         True  
2           True         True         True  
3           True         True         True  
4           True         True         True  
..           ...          ...          ...  
495         True         True         True  
496         True         True         True  
497         True         True         True  
498         True         True         True  
499         True         True         True  

[500 rows x 36 columns]

In [35]:
#We want customer Name who has maximum profit
df[df['profit']] == max(df['profit'])

KeyError: ignored

In [36]:
df['country'] == 'Sweden'

0        False
1        False
2        False
3         True
4        False
         ...  
51285    False
51286    False
51287    False
51288    False
51289    False
Name: country, Length: 51290, dtype: bool

In [37]:
  len(df[df['country'] == 'Sweden'])

203

In [38]:
df[df['shipping_cost'] > 20]

order_id  order_date   ship_date       ship_mode  \
0         AG-2011-2040    1/1/2011    1/6/2011  Standard Class   
6         CA-2011-1510    1/2/2011    1/6/2011  Standard Class   
7        IN-2011-79397    1/3/2011    1/3/2011        Same Day   
8        ID-2011-80230    1/3/2011    1/9/2011  Standard Class   
9         IZ-2011-4680    1/3/2011    1/7/2011  Standard Class   
...                ...         ...         ...             ...   
51241  ES-2014-3638865  12/31/2014    1/1/2015     First Class   
51242   MX-2014-141712  12/31/2014  12/31/2014        Same Day   
51243    IN-2014-78627  12/31/2014    1/4/2015  Standard Class   
51244  IT-2014-3715679  12/31/2014    1/5/2015    Second Class   
51245    IN-2014-75603  12/31/2014    1/5/2015    Second Class   

          customer_name      segment            state         country  market  \
0       Toby Braunhardt     Consumer      Constantine         Algeria  Africa   
6       Magdelene Morse     Consumer          Ontario          Canada  Canada   
7           Kean Nguyen    Corporate  New South Wales       Australia    APAC   
8          Ken Lonsdale     Consumer         Auckland     New Zealand    APAC   
9      Lindsay Williams    Corporate           Ninawa            Iraq    EMEA   
...                 ...          ...              ...             ...     ...   
51241   Susan Vittorini     Consumer          England  United Kingdom      EU   
51242      Rick Bensley  Home Office          Managua       Nicaragua   LATAM   
51243     Maureen Gnade     Consumer            Jilin           China    APAC   
51244          Jim Kriz  Home Office    North Brabant     Netherlands      EU   
51245      Bill Shonely    Corporate   Andhra Pradesh           India    APAC   

             region  ... sub_category                      product_name sales  \
0            Africa  ...      Storage               Tenex Lockers, Blue   408   
6            Canada  ...     Machines          Okidata Inkjet, Wireless   314   
7           Oceania  ...   Appliances           Hoover Microwave, White   276   
8           Oceania  ...      Copiers       Hewlett Wireless Fax, Laser   912   
9              EMEA  ...       Chairs  Novimex Swivel Stool, Set of Two   667   
...             ...  ...          ...                               ...   ...   
51241         North  ...      Storage   Fellowes Shelving, Single Width   289   
51242       Central  ...       Chairs      Novimex Chairmat, Set of Two    78   
51243    North Asia  ...  Accessories    Memorex Memory Card, Erganomic   300   
51244       Central  ...     Machines               StarTech Phone, Red   108   
51245  Central Asia  ...  Accessories       Memorex Keyboard, Erganomic   277   

      quantity discount   profit  shipping_cost  order_priority  year  \
0            2      0.0  106.140          35.46          Medium  2011   
6            1      0.0    3.120          24.10          Medium  2011   
7            1      0.1  110.412         125.32        Critical  2011   
8            4      0.4 -319.464         107.10             Low  2011   
9            4      0.0  253.320          81.26            High  2011   
...        ...      ...      ...            ...             ...   ...   
51241        5      0.0   75.150          34.47        Critical  2014   
51242        2      0.0   35.920          25.16        Critical  2014   
51243        3      0.0   84.060          23.96          Medium  2014   
51244        3      0.5  -19.395          22.09            High  2014   
51245        4      0.0  105.000          21.84          Medium  2014   

      category1  
0        Gaurav  
6        Gaurav  
7        Gaurav  
8        Gaurav  
9        Gaurav  
...         ...  
51241    Gaurav  
51242    Gaurav  
51243    Gaurav  
51244    Gaurav  
51245    Gaurav  

[14843 rows x 22 columns]

In [39]:
df[df['shipping_cost'] > 80]['country']

7          Australia
8        New Zealand
9               Iraq
27            Brazil
43            Mexico
            ...     
51154      Indonesia
51228         Brazil
51229      Indonesia
51230         Russia
51231       Thailand
Name: country, Length: 3984, dtype: object

In [40]:
df.head()

order_id order_date ship_date       ship_mode    customer_name  \
0     AG-2011-2040   1/1/2011  1/6/2011  Standard Class  Toby Braunhardt   
1    IN-2011-47883   1/1/2011  1/8/2011  Standard Class      Joseph Holt   
2     HU-2011-1220   1/1/2011  1/5/2011    Second Class    Annie Thurman   
3  IT-2011-3647632   1/1/2011  1/5/2011    Second Class     Eugene Moren   
4    IN-2011-47883   1/1/2011  1/8/2011  Standard Class      Joseph Holt   

       segment            state    country  market   region  ... sub_category  \
0     Consumer      Constantine    Algeria  Africa   Africa  ...      Storage   
1     Consumer  New South Wales  Australia    APAC  Oceania  ...     Supplies   
2     Consumer         Budapest    Hungary    EMEA     EMEA  ...      Storage   
3  Home Office        Stockholm     Sweden      EU    North  ...        Paper   
4     Consumer  New South Wales  Australia    APAC  Oceania  ...  Furnishings   

                  product_name sales quantity discount   profit  \
0          Tenex Lockers, Blue   408        2      0.0  106.140   
1     Acme Trimmer, High Speed   120        3      0.1   36.036   
2      Tenex Box, Single Width    66        4      0.0   29.640   
3  Enermax Note Cards, Premium    45        3      0.5  -26.055   
4   Eldon Light Bulb, Duo Pack   114        5      0.1   37.770   

   shipping_cost  order_priority  year category1  
0          35.46          Medium  2011    Gaurav  
1           9.72          Medium  2011    Gaurav  
2           8.17            High  2011    Gaurav  
3           4.82            High  2011    Gaurav  
4           4.70          Medium  2011    Gaurav  

[5 rows x 22 columns]

In [41]:
#Shipping cost > 100 and proft is less than 10
df[(df['shipping_cost'] > 100) & (df['profit'] < 10)]

order_id  order_date   ship_date       ship_mode  \
8        ID-2011-80230    1/3/2011    1/9/2011  Standard Class   
257     MX-2011-146584   1/21/2011   1/27/2011  Standard Class   
451    ES-2011-1640672    2/2/2011    2/5/2011     First Class   
625      ID-2011-47575   2/15/2011   2/21/2011  Standard Class   
626      ID-2011-47575   2/15/2011   2/21/2011  Standard Class   
...                ...         ...         ...             ...   
50781   CA-2014-147956  12/25/2014    1/1/2015  Standard Class   
50969   US-2014-120166  12/27/2014  12/31/2014  Standard Class   
50970  ES-2014-4296634  12/27/2014    1/2/2015  Standard Class   
51034    ID-2014-68183  12/29/2014  12/31/2014     First Class   
51153    ID-2014-51509  12/30/2014    1/1/2015    Second Class   

              customer_name    segment                         state  \
8              Ken Lonsdale   Consumer                      Auckland   
257            Hunter Lopez   Consumer                       Managua   
451              Ken Heidel  Corporate                        Saxony   
625    Christopher Martinez   Consumer              National Capital   
626    Christopher Martinez   Consumer              National Capital   
...                     ...        ...                           ...   
50781            Alan Hwang   Consumer                    Washington   
50969          Frank Hawley  Corporate               Valle del Cauca   
50970           Harry Marie  Corporate                      Brittany   
51034             Adam Hart  Corporate                      Victoria   
51153          Naresj Patel   Consumer  Australian Capital Territory   

             country market          region  ... sub_category  \
8        New Zealand   APAC         Oceania  ...      Copiers   
257        Nicaragua  LATAM         Central  ...       Tables   
451          Germany     EU         Central  ...       Phones   
625      Philippines   APAC  Southeast Asia  ...       Phones   
626      Philippines   APAC  Southeast Asia  ...      Copiers   
...              ...    ...             ...  ...          ...   
50781  United States     US            West  ...      Storage   
50969       Colombia  LATAM           South  ...       Tables   
50970         France     EU         Central  ...   Appliances   
51034      Australia   APAC         Oceania  ...   Appliances   
51153      Australia   APAC         Oceania  ...       Chairs   

                                            product_name  sales quantity  \
8                            Hewlett Wireless Fax, Laser    912        4   
257         Chromcraft Conference Table, Fully Assembled  2,797        6   
451                    Motorola Signal Booster, Cordless    584        8   
625                 Nokia Signal Booster, with Caller ID    921        9   
626                      Brother Fax Machine, High-Speed  1,235        6   
...                                                  ...    ...      ...   
50781  Office Impressions Heavy Duty Welded Shelving ...  1,004        6   
50969               Barricks Training Table, Rectangular    842        5   
50970                    Hamilton Beach Microwave, White  1,010        4   
51034                                  Hoover Stove, Red  1,535        3   
51153             Harbour Creations Rocking Chair, Black    434        5   

      discount   profit  shipping_cost  order_priority  year category1  
8         0.40 -319.464         107.10             Low  2011    Gaurav  
257       0.20   -0.072         320.38             Low  2011    Gaurav  
451       0.50  -58.560         109.14            High  2011    Gaurav  
625       0.25 -172.125         128.01             Low  2011    Gaurav  
626       0.35 -456.174         118.63             Low  2011    Gaurav  
...        ...      ...            ...             ...   ...       ...  
50781     0.00    0.000         121.83             Low  2014    Gaurav  
50969     0.20 -200.080         120.11            High  2014    Gaurav  
50970

In [42]:
df.dtypes #Date is being treated as String/Object

order_id           object
order_date         object
ship_date          object
ship_mode          object
customer_name      object
segment            object
state              object
country            object
market             object
region             object
product_id         object
category           object
sub_category       object
product_name       object
sales              object
quantity            int64
discount          float64
profit            float64
shipping_cost     float64
order_priority     object
year                int64
category1          object
dtype: object

In [43]:
#Converting into Date type Format
pd.to_datetime(df['order_date'])

0       2011-01-01
1       2011-01-01
2       2011-01-01
3       2011-01-01
4       2011-01-01
           ...    
51285   2014-12-31
51286   2014-12-31
51287   2014-12-31
51288   2014-12-31
51289   2014-12-31
Name: order_date, Length: 51290, dtype: datetime64[ns]

In [44]:
date = pd.to_datetime(df['order_date'])
date

0       2011-01-01
1       2011-01-01
2       2011-01-01
3       2011-01-01
4       2011-01-01
           ...    
51285   2014-12-31
51286   2014-12-31
51287   2014-12-31
51288   2014-12-31
51289   2014-12-31
Name: order_date, Length: 51290, dtype: datetime64[ns]

In [45]:
type(date[0])

pandas._libs.tslibs.timestamps.Timestamp

In [46]:
  df['converted_order_date'] = pd.to_datetime(df['order_date'])
  df

order_id  order_date ship_date       ship_mode  \
0         AG-2011-2040    1/1/2011  1/6/2011  Standard Class   
1        IN-2011-47883    1/1/2011  1/8/2011  Standard Class   
2         HU-2011-1220    1/1/2011  1/5/2011    Second Class   
3      IT-2011-3647632    1/1/2011  1/5/2011    Second Class   
4        IN-2011-47883    1/1/2011  1/8/2011  Standard Class   
...                ...         ...       ...             ...   
51285   CA-2014-115427  12/31/2014  1/4/2015  Standard Class   
51286     MO-2014-2560  12/31/2014  1/5/2015  Standard Class   
51287   MX-2014-110527  12/31/2014  1/2/2015    Second Class   
51288   MX-2014-114783  12/31/2014  1/6/2015  Standard Class   
51289   CA-2014-156720  12/31/2014  1/4/2015  Standard Class   

          customer_name      segment             state        country  market  \
0       Toby Braunhardt     Consumer       Constantine        Algeria  Africa   
1           Joseph Holt     Consumer   New South Wales      Australia    APAC   
2         Annie Thurman     Consumer          Budapest        Hungary    EMEA   
3          Eugene Moren  Home Office         Stockholm         Sweden      EU   
4           Joseph Holt     Consumer   New South Wales      Australia    APAC   
...                 ...          ...               ...            ...     ...   
51285        Erica Bern    Corporate        California  United States      US   
51286         Liz Preis     Consumer  Souss-Massa-Draƒ        Morocco  Africa   
51287  Charlotte Melton     Consumer           Managua      Nicaragua   LATAM   
51288     Tamara Dahlen     Consumer         Chihuahua         Mexico   LATAM   
51289     Jill Matthias     Consumer          Colorado  United States      US   

        region  ...                                     product_name sales  \
0       Africa  ...                              Tenex Lockers, Blue   408   
1      Oceania  ...                         Acme Trimmer, High Speed   120   
2         EMEA  ...                          Tenex Box, Single Width    66   
3        North  ...                      Enermax Note Cards, Premium    45   
4      Oceania  ...                       Eldon Light Bulb, Duo Pack   114   
...        ...  ...                                              ...   ...   
51285     West  ...  Cardinal Slant-D Ring Binder, Heavy Gauge Vinyl    14   
51286   Africa  ...          Wilson Jones Hole Reinforcements, Clear     4   
51287  Central  ...           Hon Color Coded Labels, 5000 Label Set    26   
51288    North  ...           Hon Legal Exhibit Labels, Alphabetical     7   
51289     West  ...                              Bagged Rubber Bands     3   

      quantity discount    profit  shipping_cost  order_priority  year  \
0            2      0.0  106.1400          35.46          Medium  2011   
1            3      0.1   36.0360           9.72          Medium  2011   
2            4      0.0   29.6400           8.17            High  2011   
3            3      0.5  -26.0550           4.82            High  2011   
4            5      0.1   37.7700           4.70          Medium  2011   
...        ...      ...       ...            ...             ...   ...   
51285        2      0.2    4.5188           0.89          Medium  2014   
51286        1      0.0    0.4200           0.49          Medium  2014   
51287        3      0.0   12.3600           0.35          Medium  2014   
51288        1      0.0    0.5600           0.20          Medium  2014   
51289        3      0.2   -0.6048           0.17          Medium  2014   

       category1 converted_order_date  
0         Gaurav           2011-01-01  
1         Gaurav           2011-01-01  
2         Gaurav           2011-01-01  
3         Gaurav           2011-01-01  
4         Gaurav           2011-01-01  
...          ...                  ...  
51285     Gaurav           2014-12-31  
51286     Gaurav           2014-12-31  
51287     Gaurav           2014-12-31  
51288     Gaurav           2014-12-31  
51289 

In [47]:
#Extracting only year from the date
df['order_date_year'] = df['converted_order_date'].dt.year
df

order_id  order_date ship_date       ship_mode  \
0         AG-2011-2040    1/1/2011  1/6/2011  Standard Class   
1        IN-2011-47883    1/1/2011  1/8/2011  Standard Class   
2         HU-2011-1220    1/1/2011  1/5/2011    Second Class   
3      IT-2011-3647632    1/1/2011  1/5/2011    Second Class   
4        IN-2011-47883    1/1/2011  1/8/2011  Standard Class   
...                ...         ...       ...             ...   
51285   CA-2014-115427  12/31/2014  1/4/2015  Standard Class   
51286     MO-2014-2560  12/31/2014  1/5/2015  Standard Class   
51287   MX-2014-110527  12/31/2014  1/2/2015    Second Class   
51288   MX-2014-114783  12/31/2014  1/6/2015  Standard Class   
51289   CA-2014-156720  12/31/2014  1/4/2015  Standard Class   

          customer_name      segment             state        country  market  \
0       Toby Braunhardt     Consumer       Constantine        Algeria  Africa   
1           Joseph Holt     Consumer   New South Wales      Australia    APAC   
2         Annie Thurman     Consumer          Budapest        Hungary    EMEA   
3          Eugene Moren  Home Office         Stockholm         Sweden      EU   
4           Joseph Holt     Consumer   New South Wales      Australia    APAC   
...                 ...          ...               ...            ...     ...   
51285        Erica Bern    Corporate        California  United States      US   
51286         Liz Preis     Consumer  Souss-Massa-Draƒ        Morocco  Africa   
51287  Charlotte Melton     Consumer           Managua      Nicaragua   LATAM   
51288     Tamara Dahlen     Consumer         Chihuahua         Mexico   LATAM   
51289     Jill Matthias     Consumer          Colorado  United States      US   

        region  ... sales quantity discount    profit shipping_cost  \
0       Africa  ...   408        2      0.0  106.1400         35.46   
1      Oceania  ...   120        3      0.1   36.0360          9.72   
2         EMEA  ...    66        4      0.0   29.6400          8.17   
3        North  ...    45        3      0.5  -26.0550          4.82   
4      Oceania  ...   114        5      0.1   37.7700          4.70   
...        ...  ...   ...      ...      ...       ...           ...   
51285     West  ...    14        2      0.2    4.5188          0.89   
51286   Africa  ...     4        1      0.0    0.4200          0.49   
51287  Central  ...    26        3      0.0   12.3600          0.35   
51288    North  ...     7        1      0.0    0.5600          0.20   
51289     West  ...     3        3      0.2   -0.6048          0.17   

       order_priority  year  category1  converted_order_date order_date_year  
0              Medium  2011     Gaurav            2011-01-01            2011  
1              Medium  2011     Gaurav            2011-01-01            2011  
2                High  2011     Gaurav            2011-01-01            2011  
3                High  2011     Gaurav            2011-01-01            2011  
4              Medium  2011     Gaurav            2011-01-01            2011  
...               ...   ...        ...                   ...             ...  
51285          Medium  2014     Gaurav            2014-12-31            2014  
51286          Medium  2014     Gaurav            2014-12-31            2014  
51287          Medium  2014     Gaurav            2014-12-31            2014  
51288          Medium  2014     Gaurav            2014-12-31            2014  
51289          Medium  2014     Gaurav            2014-12-31            2014  

[51290 rows x 24 columns]

In [48]:
#Extracting only month from the date
df['order_date_month'] = df['converted_order_date'].dt.month
df

order_id  order_date ship_date       ship_mode  \
0         AG-2011-2040    1/1/2011  1/6/2011  Standard Class   
1        IN-2011-47883    1/1/2011  1/8/2011  Standard Class   
2         HU-2011-1220    1/1/2011  1/5/2011    Second Class   
3      IT-2011-3647632    1/1/2011  1/5/2011    Second Class   
4        IN-2011-47883    1/1/2011  1/8/2011  Standard Class   
...                ...         ...       ...             ...   
51285   CA-2014-115427  12/31/2014  1/4/2015  Standard Class   
51286     MO-2014-2560  12/31/2014  1/5/2015  Standard Class   
51287   MX-2014-110527  12/31/2014  1/2/2015    Second Class   
51288   MX-2014-114783  12/31/2014  1/6/2015  Standard Class   
51289   CA-2014-156720  12/31/2014  1/4/2015  Standard Class   

          customer_name      segment             state        country  market  \
0       Toby Braunhardt     Consumer       Constantine        Algeria  Africa   
1           Joseph Holt     Consumer   New South Wales      Australia    APAC   
2         Annie Thurman     Consumer          Budapest        Hungary    EMEA   
3          Eugene Moren  Home Office         Stockholm         Sweden      EU   
4           Joseph Holt     Consumer   New South Wales      Australia    APAC   
...                 ...          ...               ...            ...     ...   
51285        Erica Bern    Corporate        California  United States      US   
51286         Liz Preis     Consumer  Souss-Massa-Draƒ        Morocco  Africa   
51287  Charlotte Melton     Consumer           Managua      Nicaragua   LATAM   
51288     Tamara Dahlen     Consumer         Chihuahua         Mexico   LATAM   
51289     Jill Matthias     Consumer          Colorado  United States      US   

        region  ... quantity discount    profit shipping_cost order_priority  \
0       Africa  ...        2      0.0  106.1400         35.46         Medium   
1      Oceania  ...        3      0.1   36.0360          9.72         Medium   
2         EMEA  ...        4      0.0   29.6400          8.17           High   
3        North  ...        3      0.5  -26.0550          4.82           High   
4      Oceania  ...        5      0.1   37.7700          4.70         Medium   
...        ...  ...      ...      ...       ...           ...            ...   
51285     West  ...        2      0.2    4.5188          0.89         Medium   
51286   Africa  ...        1      0.0    0.4200          0.49         Medium   
51287  Central  ...        3      0.0   12.3600          0.35         Medium   
51288    North  ...        1      0.0    0.5600          0.20         Medium   
51289     West  ...        3      0.2   -0.6048          0.17         Medium   

       year  category1  converted_order_date  order_date_year order_date_month  
0      2011     Gaurav            2011-01-01             2011                1  
1      2011     Gaurav            2011-01-01             2011                1  
2      2011     Gaurav            2011-01-01             2011                1  
3      2011     Gaurav            2011-01-01             2011                1  
4      2011     Gaurav            2011-01-01             2011                1  
...     ...        ...                   ...              ...              ...  
51285  2014     Gaurav            2014-12-31             2014               12  
51286  2014     Gaurav            2014-12-31             2014               12  
51287  2014     Gaurav            2014-12-31             2014               12  
51288  2014     Gaurav            2014-12-31             2014               12  
51289  2014     Gaurav            2014-12-31             2014               12  

[51290 rows x 25 columns]

In [49]:
#Extracting week of the year from the date
df['order_date_week'] = df['converted_order_date'].dt.week
df

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: FutureWarning: Series.dt.weekofyear and Series.dt.week have been deprecated.  Please use Series.dt.isocalendar().week instead.
  


order_id  order_date ship_date       ship_mode  \
0         AG-2011-2040    1/1/2011  1/6/2011  Standard Class   
1        IN-2011-47883    1/1/2011  1/8/2011  Standard Class   
2         HU-2011-1220    1/1/2011  1/5/2011    Second Class   
3      IT-2011-3647632    1/1/2011  1/5/2011    Second Class   
4        IN-2011-47883    1/1/2011  1/8/2011  Standard Class   
...                ...         ...       ...             ...   
51285   CA-2014-115427  12/31/2014  1/4/2015  Standard Class   
51286     MO-2014-2560  12/31/2014  1/5/2015  Standard Class   
51287   MX-2014-110527  12/31/2014  1/2/2015    Second Class   
51288   MX-2014-114783  12/31/2014  1/6/2015  Standard Class   
51289   CA-2014-156720  12/31/2014  1/4/2015  Standard Class   

          customer_name      segment             state        country  market  \
0       Toby Braunhardt     Consumer       Constantine        Algeria  Africa   
1           Joseph Holt     Consumer   New South Wales      Australia    APAC   
2         Annie Thurman     Consumer          Budapest        Hungary    EMEA   
3          Eugene Moren  Home Office         Stockholm         Sweden      EU   
4           Joseph Holt     Consumer   New South Wales      Australia    APAC   
...                 ...          ...               ...            ...     ...   
51285        Erica Bern    Corporate        California  United States      US   
51286         Liz Preis     Consumer  Souss-Massa-Draƒ        Morocco  Africa   
51287  Charlotte Melton     Consumer           Managua      Nicaragua   LATAM   
51288     Tamara Dahlen     Consumer         Chihuahua         Mexico   LATAM   
51289     Jill Matthias     Consumer          Colorado  United States      US   

        region  ... discount    profit shipping_cost order_priority  year  \
0       Africa  ...      0.0  106.1400         35.46         Medium  2011   
1      Oceania  ...      0.1   36.0360          9.72         Medium  2011   
2         EMEA  ...      0.0   29.6400          8.17           High  2011   
3        North  ...      0.5  -26.0550          4.82           High  2011   
4      Oceania  ...      0.1   37.7700          4.70         Medium  2011   
...        ...  ...      ...       ...           ...            ...   ...   
51285     West  ...      0.2    4.5188          0.89         Medium  2014   
51286   Africa  ...      0.0    0.4200          0.49         Medium  2014   
51287  Central  ...      0.0   12.3600          0.35         Medium  2014   
51288    North  ...      0.0    0.5600          0.20         Medium  2014   
51289     West  ...      0.2   -0.6048          0.17         Medium  2014   

       category1  converted_order_date  order_date_year  order_date_month  \
0         Gaurav            2011-01-01             2011                 1   
1         Gaurav            2011-01-01             2011                 1   
2         Gaurav            2011-01-01             2011                 1   
3         Gaurav            2011-01-01             2011                 1   
4         Gaurav            2011-01-01             2011                 1   
...          ...                   ...              ...               ...   
51285     Gaurav            2014-12-31             2014                12   
51286     Gaurav            2014-12-31             2014                12   
51287     Gaurav            2014-12-31             2014                12   
51288     Gaurav            2014-12-31             2014                12   
51289     Gaurav            2014-12-31             2014                12   

      order_date_week  
0                  52  
1                  52  
2                  52  
3                  52  
4                  52  
...               ...  
51285               1  
51286               1  
51287               1  
51288               1  
51289               1  

[51290 rows x 26 columns]

In [50]:
df.columns

Index(['order_id', 'order_date', 'ship_date', 'ship_mode', 'customer_name',
       'segment', 'state', 'country', 'market', 'region', 'product_id',
       'category', 'sub_category', 'product_name', 'sales', 'quantity',
       'discount', 'profit', 'shipping_cost', 'order_priority', 'year',
       'category1', 'converted_order_date', 'order_date_year',
       'order_date_month', 'order_date_week'],
      dtype='object')

In [51]:
df[['converted_order_date','order_date_week']] #Showing jan as 52nd week and December as 1st Week. 
#So basically it shows how much week is left for the year

converted_order_date  order_date_week
0               2011-01-01               52
1               2011-01-01               52
2               2011-01-01               52
3               2011-01-01               52
4               2011-01-01               52
...                    ...              ...
51285           2014-12-31                1
51286           2014-12-31                1
51287           2014-12-31                1
51288           2014-12-31                1
51289           2014-12-31                1

[51290 rows x 2 columns]

In [52]:
#In which month, we are going to do most sales --> Group by operation on month and then count
#In Pandas, this is equivalent to value_counts function
df['order_date_month'].value_counts()

11    6342
12    6302
9     5975
6     5331
8     5078
10    4490
5     3747
7     3166
4     3057
3     3035
1     2599
2     2168
Name: order_date_month, dtype: int64

In [53]:
df.columns

Index(['order_id', 'order_date', 'ship_date', 'ship_mode', 'customer_name',
       'segment', 'state', 'country', 'market', 'region', 'product_id',
       'category', 'sub_category', 'product_name', 'sales', 'quantity',
       'discount', 'profit', 'shipping_cost', 'order_priority', 'year',
       'category1', 'converted_order_date', 'order_date_year',
       'order_date_month', 'order_date_week'],
      dtype='object')

In [54]:
#Cost to Company Computation
df['cost_to_company'] = df['discount'] + df['shipping_cost']
df.head()

order_id order_date ship_date       ship_mode    customer_name  \
0     AG-2011-2040   1/1/2011  1/6/2011  Standard Class  Toby Braunhardt   
1    IN-2011-47883   1/1/2011  1/8/2011  Standard Class      Joseph Holt   
2     HU-2011-1220   1/1/2011  1/5/2011    Second Class    Annie Thurman   
3  IT-2011-3647632   1/1/2011  1/5/2011    Second Class     Eugene Moren   
4    IN-2011-47883   1/1/2011  1/8/2011  Standard Class      Joseph Holt   

       segment            state    country  market   region  ...   profit  \
0     Consumer      Constantine    Algeria  Africa   Africa  ...  106.140   
1     Consumer  New South Wales  Australia    APAC  Oceania  ...   36.036   
2     Consumer         Budapest    Hungary    EMEA     EMEA  ...   29.640   
3  Home Office        Stockholm     Sweden      EU    North  ...  -26.055   
4     Consumer  New South Wales  Australia    APAC  Oceania  ...   37.770   

  shipping_cost order_priority  year category1  converted_order_date  \
0         35.46         Medium  2011    Gaurav            2011-01-01   
1          9.72         Medium  2011    Gaurav            2011-01-01   
2          8.17           High  2011    Gaurav            2011-01-01   
3          4.82           High  2011    Gaurav            2011-01-01   
4          4.70         Medium  2011    Gaurav            2011-01-01   

   order_date_year  order_date_month  order_date_week cost_to_company  
0             2011                 1               52           35.46  
1             2011                 1               52            9.82  
2             2011                 1               52            8.17  
3             2011                 1               52            5.32  
4             2011                 1               52            4.80  

[5 rows x 27 columns]

In [55]:
df[df['cost_to_company'] == max(df['cost_to_company'])]['product_name']

13891    Plantronics CS510 - Over-the-Head monaural Wir...
Name: product_name, dtype: object

In [ ]:
#Dropping Column
df.drop('cost_to_company',axis=1,inplace=True) #Column deleted permanently because of inplace=True, axis = 1 means column operation

In [59]:
df.drop(1) #Delete row 1 with index. Still we have 1st row when we display df
df

order_id  order_date ship_date       ship_mode  \
0         AG-2011-2040    1/1/2011  1/6/2011  Standard Class   
1        IN-2011-47883    1/1/2011  1/8/2011  Standard Class   
2         HU-2011-1220    1/1/2011  1/5/2011    Second Class   
3      IT-2011-3647632    1/1/2011  1/5/2011    Second Class   
4        IN-2011-47883    1/1/2011  1/8/2011  Standard Class   
...                ...         ...       ...             ...   
51285   CA-2014-115427  12/31/2014  1/4/2015  Standard Class   
51286     MO-2014-2560  12/31/2014  1/5/2015  Standard Class   
51287   MX-2014-110527  12/31/2014  1/2/2015    Second Class   
51288   MX-2014-114783  12/31/2014  1/6/2015  Standard Class   
51289   CA-2014-156720  12/31/2014  1/4/2015  Standard Class   

          customer_name      segment             state        country  market  \
0       Toby Braunhardt     Consumer       Constantine        Algeria  Africa   
1           Joseph Holt     Consumer   New South Wales      Australia    APAC   
2         Annie Thurman     Consumer          Budapest        Hungary    EMEA   
3          Eugene Moren  Home Office         Stockholm         Sweden      EU   
4           Joseph Holt     Consumer   New South Wales      Australia    APAC   
...                 ...          ...               ...            ...     ...   
51285        Erica Bern    Corporate        California  United States      US   
51286         Liz Preis     Consumer  Souss-Massa-Draƒ        Morocco  Africa   
51287  Charlotte Melton     Consumer           Managua      Nicaragua   LATAM   
51288     Tamara Dahlen     Consumer         Chihuahua         Mexico   LATAM   
51289     Jill Matthias     Consumer          Colorado  United States      US   

        region  ... discount    profit shipping_cost order_priority  year  \
0       Africa  ...      0.0  106.1400         35.46         Medium  2011   
1      Oceania  ...      0.1   36.0360          9.72         Medium  2011   
2         EMEA  ...      0.0   29.6400          8.17           High  2011   
3        North  ...      0.5  -26.0550          4.82           High  2011   
4      Oceania  ...      0.1   37.7700          4.70         Medium  2011   
...        ...  ...      ...       ...           ...            ...   ...   
51285     West  ...      0.2    4.5188          0.89         Medium  2014   
51286   Africa  ...      0.0    0.4200          0.49         Medium  2014   
51287  Central  ...      0.0   12.3600          0.35         Medium  2014   
51288    North  ...      0.0    0.5600          0.20         Medium  2014   
51289     West  ...      0.2   -0.6048          0.17         Medium  2014   

       category1  converted_order_date  order_date_year  order_date_month  \
0         Gaurav            2011-01-01             2011                 1   
1         Gaurav            2011-01-01             2011                 1   
2         Gaurav            2011-01-01             2011                 1   
3         Gaurav            2011-01-01             2011                 1   
4         Gaurav            2011-01-01             2011                 1   
...          ...                   ...              ...               ...   
51285     Gaurav            2014-12-31             2014                12   
51286     Gaurav            2014-12-31             2014                12   
51287     Gaurav            2014-12-31             2014                12   
51288     Gaurav            2014-12-31             2014                12   
51289     Gaurav            2014-12-31             2014                12   

      order_date_week  
0                  52  
1                  52  
2                  52  
3                  52  
4                  52  
...               ...  
51285               1  
51286               1  
51287               1  
51288               1  
51289               1  

[51290 rows x 26 columns]

In [61]:
#To drop 1st row permanently
df.drop(1,inplace=True)
df

order_id  order_date ship_date       ship_mode  \
0         AG-2011-2040    1/1/2011  1/6/2011  Standard Class   
2         HU-2011-1220    1/1/2011  1/5/2011    Second Class   
3      IT-2011-3647632    1/1/2011  1/5/2011    Second Class   
4        IN-2011-47883    1/1/2011  1/8/2011  Standard Class   
5        IN-2011-47883    1/1/2011  1/8/2011  Standard Class   
...                ...         ...       ...             ...   
51285   CA-2014-115427  12/31/2014  1/4/2015  Standard Class   
51286     MO-2014-2560  12/31/2014  1/5/2015  Standard Class   
51287   MX-2014-110527  12/31/2014  1/2/2015    Second Class   
51288   MX-2014-114783  12/31/2014  1/6/2015  Standard Class   
51289   CA-2014-156720  12/31/2014  1/4/2015  Standard Class   

          customer_name      segment             state        country  market  \
0       Toby Braunhardt     Consumer       Constantine        Algeria  Africa   
2         Annie Thurman     Consumer          Budapest        Hungary    EMEA   
3          Eugene Moren  Home Office         Stockholm         Sweden      EU   
4           Joseph Holt     Consumer   New South Wales      Australia    APAC   
5           Joseph Holt     Consumer   New South Wales      Australia    APAC   
...                 ...          ...               ...            ...     ...   
51285        Erica Bern    Corporate        California  United States      US   
51286         Liz Preis     Consumer  Souss-Massa-Draƒ        Morocco  Africa   
51287  Charlotte Melton     Consumer           Managua      Nicaragua   LATAM   
51288     Tamara Dahlen     Consumer         Chihuahua         Mexico   LATAM   
51289     Jill Matthias     Consumer          Colorado  United States      US   

        region  ... discount    profit shipping_cost order_priority  year  \
0       Africa  ...      0.0  106.1400         35.46         Medium  2011   
2         EMEA  ...      0.0   29.6400          8.17           High  2011   
3        North  ...      0.5  -26.0550          4.82           High  2011   
4      Oceania  ...      0.1   37.7700          4.70         Medium  2011   
5      Oceania  ...      0.1   15.3420          1.80         Medium  2011   
...        ...  ...      ...       ...           ...            ...   ...   
51285     West  ...      0.2    4.5188          0.89         Medium  2014   
51286   Africa  ...      0.0    0.4200          0.49         Medium  2014   
51287  Central  ...      0.0   12.3600          0.35         Medium  2014   
51288    North  ...      0.0    0.5600          0.20         Medium  2014   
51289     West  ...      0.2   -0.6048          0.17         Medium  2014   

       category1  converted_order_date  order_date_year  order_date_month  \
0         Gaurav            2011-01-01             2011                 1   
2         Gaurav            2011-01-01             2011                 1   
3         Gaurav            2011-01-01             2011                 1   
4         Gaurav            2011-01-01             2011                 1   
5         Gaurav            2011-01-01             2011                 1   
...          ...                   ...              ...               ...   
51285     Gaurav            2014-12-31             2014                12   
51286     Gaurav            2014-12-31             2014                12   
51287     Gaurav            2014-12-31             2014                12   
51288     Gaurav            2014-12-31             2014                12   
51289     Gaurav            2014-12-31             2014                12   

      order_date_week  
0                  52  
2                  52  
3                  52  
4                  52  
5                  52  
...               ...  
51285               1  
51286               1  
51287               1  
51288               1  
51289               1  

[51289 rows x 26 columns]

In [62]:
#Filtering Rows: loc (Location) and iloc functions
#In loc function we pass named indexes and not exact internal indexes
#In iloc function we pass row and column exact indexes
df.loc[[2,3]] #Row indexes given

order_id order_date ship_date     ship_mode  customer_name  \
2     HU-2011-1220   1/1/2011  1/5/2011  Second Class  Annie Thurman   
3  IT-2011-3647632   1/1/2011  1/5/2011  Second Class   Eugene Moren   

       segment      state  country market region  ... discount  profit  \
2     Consumer   Budapest  Hungary   EMEA   EMEA  ...      0.0  29.640   
3  Home Office  Stockholm   Sweden     EU  North  ...      0.5 -26.055   

  shipping_cost order_priority  year  category1  converted_order_date  \
2          8.17           High  2011     Gaurav            2011-01-01   
3          4.82           High  2011     Gaurav            2011-01-01   

   order_date_year  order_date_month order_date_week  
2             2011                 1              52  
3             2011                 1              52  

[2 rows x 26 columns]

In [64]:
df.loc[2:10] 

order_id order_date ship_date       ship_mode     customer_name  \
2      HU-2011-1220   1/1/2011  1/5/2011    Second Class     Annie Thurman   
3   IT-2011-3647632   1/1/2011  1/5/2011    Second Class      Eugene Moren   
4     IN-2011-47883   1/1/2011  1/8/2011  Standard Class       Joseph Holt   
5     IN-2011-47883   1/1/2011  1/8/2011  Standard Class       Joseph Holt   
6      CA-2011-1510   1/2/2011  1/6/2011  Standard Class   Magdelene Morse   
7     IN-2011-79397   1/3/2011  1/3/2011        Same Day       Kean Nguyen   
8     ID-2011-80230   1/3/2011  1/9/2011  Standard Class      Ken Lonsdale   
9      IZ-2011-4680   1/3/2011  1/7/2011  Standard Class  Lindsay Williams   
10    IN-2011-65159   1/3/2011  1/7/2011    Second Class      Larry Blacks   

        segment             state      country  market          region  ...  \
2      Consumer          Budapest      Hungary    EMEA            EMEA  ...   
3   Home Office         Stockholm       Sweden      EU           North  ...   
4      Consumer   New South Wales    Australia    APAC         Oceania  ...   
5      Consumer   New South Wales    Australia    APAC         Oceania  ...   
6      Consumer           Ontario       Canada  Canada          Canada  ...   
7     Corporate   New South Wales    Australia    APAC         Oceania  ...   
8      Consumer          Auckland  New Zealand    APAC         Oceania  ...   
9     Corporate            Ninawa         Iraq    EMEA            EMEA  ...   
10     Consumer  National Capital  Philippines    APAC  Southeast Asia  ...   

   discount    profit shipping_cost order_priority  year  category1  \
2      0.00   29.6400          8.17           High  2011     Gaurav   
3      0.50  -26.0550          4.82           High  2011     Gaurav   
4      0.10   37.7700          4.70         Medium  2011     Gaurav   
5      0.10   15.3420          1.80         Medium  2011     Gaurav   
6      0.00    3.1200         24.10         Medium  2011     Gaurav   
7      0.10  110.4120        125.32       Critical  2011     Gaurav   
8      0.40 -319.4640        107.10            Low  2011     Gaurav   
9      0.00  253.3200         81.26           High  2011     Gaurav   
10     0.45 -122.8005         33.75           High  2011     Gaurav   

    converted_order_date  order_date_year  order_date_month order_date_week  
2             2011-01-01             2011                 1              52  
3             2011-01-01             2011                 1              52  
4             2011-01-01             2011                 1              52  
5             2011-01-01             2011                 1              52  
6             2011-01-02             2011                 1              52  
7             2011-01-03             2011                 1               1  
8             2011-01-03             2011                 1               1  
9             2011-01-03             2011                 1               1  
10            2011-01-03             2011                 1               1  

[9 rows x 26 columns]

In [65]:
df.loc[2:10:2] 

order_id order_date ship_date       ship_mode    customer_name  \
2    HU-2011-1220   1/1/2011  1/5/2011    Second Class    Annie Thurman   
4   IN-2011-47883   1/1/2011  1/8/2011  Standard Class      Joseph Holt   
6    CA-2011-1510   1/2/2011  1/6/2011  Standard Class  Magdelene Morse   
8   ID-2011-80230   1/3/2011  1/9/2011  Standard Class     Ken Lonsdale   
10  IN-2011-65159   1/3/2011  1/7/2011    Second Class     Larry Blacks   

     segment             state      country  market          region  ...  \
2   Consumer          Budapest      Hungary    EMEA            EMEA  ...   
4   Consumer   New South Wales    Australia    APAC         Oceania  ...   
6   Consumer           Ontario       Canada  Canada          Canada  ...   
8   Consumer          Auckland  New Zealand    APAC         Oceania  ...   
10  Consumer  National Capital  Philippines    APAC  Southeast Asia  ...   

   discount    profit shipping_cost order_priority  year  category1  \
2      0.00   29.6400          8.17           High  2011     Gaurav   
4      0.10   37.7700          4.70         Medium  2011     Gaurav   
6      0.00    3.1200         24.10         Medium  2011     Gaurav   
8      0.40 -319.4640        107.10            Low  2011     Gaurav   
10     0.45 -122.8005         33.75           High  2011     Gaurav   

    converted_order_date  order_date_year  order_date_month order_date_week  
2             2011-01-01             2011                 1              52  
4             2011-01-01             2011                 1              52  
6             2011-01-02             2011                 1              52  
8             2011-01-03             2011                 1               1  
10            2011-01-03             2011                 1               1  

[5 rows x 26 columns]

In [66]:
#Extracting certain row and column
df.loc[0:4,['order_id','order_date','ship_date']]

order_id order_date ship_date
0     AG-2011-2040   1/1/2011  1/6/2011
2     HU-2011-1220   1/1/2011  1/5/2011
3  IT-2011-3647632   1/1/2011  1/5/2011
4    IN-2011-47883   1/1/2011  1/8/2011

In [67]:
#iloc function: Here we give row and column default (internal) index instead of named indexes
df.loc[0:4,0:3] #This will not work as we have to specify column named indexes and not default indexes
df.iloc[0:4,0:3]

order_id order_date ship_date
0     AG-2011-2040   1/1/2011  1/6/2011
2     HU-2011-1220   1/1/2011  1/5/2011
3  IT-2011-3647632   1/1/2011  1/5/2011
4    IN-2011-47883   1/1/2011  1/8/2011

In [71]:
df.iloc[1:6,3:6]

ship_mode    customer_name      segment
2    Second Class    Annie Thurman     Consumer
3    Second Class     Eugene Moren  Home Office
4  Standard Class      Joseph Holt     Consumer
5  Standard Class      Joseph Holt     Consumer
6  Standard Class  Magdelene Morse     Consumer

In [75]:
#Performing Filter Options
df.dtypes[(df.dtypes == 'float64') | (df.dtypes == 'int64')]

quantity              int64
discount            float64
profit              float64
shipping_cost       float64
year                  int64
order_date_year       int64
order_date_month      int64
order_date_week       int64
dtype: object

In [79]:
df2 = df[df.dtypes[(df.dtypes == 'float64' )| (df.dtypes == 'int64')].index]
df2

quantity  discount    profit  shipping_cost  year  order_date_year  \
0             2       0.0  106.1400          35.46  2011             2011   
2             4       0.0   29.6400           8.17  2011             2011   
3             3       0.5  -26.0550           4.82  2011             2011   
4             5       0.1   37.7700           4.70  2011             2011   
5             2       0.1   15.3420           1.80  2011             2011   
...         ...       ...       ...            ...   ...              ...   
51285         2       0.2    4.5188           0.89  2014             2014   
51286         1       0.0    0.4200           0.49  2014             2014   
51287         3       0.0   12.3600           0.35  2014             2014   
51288         1       0.0    0.5600           0.20  2014             2014   
51289         3       0.2   -0.6048           0.17  2014             2014   

       order_date_month  order_date_week  
0                     1               52  
2                     1               52  
3                     1               52  
4                     1               52  
5                     1               52  
...                 ...              ...  
51285                12                1  
51286                12                1  
51287                12                1  
51288                12                1  
51289                12                1  

[51289 rows x 8 columns]

In [80]:
df2[df2>100]

quantity  discount  profit  shipping_cost  year  order_date_year  \
0           NaN       NaN  106.14            NaN  2011             2011   
2           NaN       NaN     NaN            NaN  2011             2011   
3           NaN       NaN     NaN            NaN  2011             2011   
4           NaN       NaN     NaN            NaN  2011             2011   
5           NaN       NaN     NaN            NaN  2011             2011   
...         ...       ...     ...            ...   ...              ...   
51285       NaN       NaN     NaN            NaN  2014             2014   
51286       NaN       NaN     NaN            NaN  2014             2014   
51287       NaN       NaN     NaN            NaN  2014             2014   
51288       NaN       NaN     NaN            NaN  2014             2014   
51289       NaN       NaN     NaN            NaN  2014             2014   

       order_date_month  order_date_week  
0                   NaN              NaN  
2                   NaN              NaN  
3                   NaN              NaN  
4                   NaN              NaN  
5                   NaN              NaN  
...                 ...              ...  
51285               NaN              NaN  
51286               NaN              NaN  
51287               NaN              NaN  
51288               NaN              NaN  
51289               NaN              NaN  

[51289 rows x 8 columns]

In [95]:
df3 = df2[df2 >100]

In [96]:
#Will drop if there is a single NA value in a row
df3.dropna() #If we want permanent delete then inplace = True needs to be used

Empty DataFrame
Columns: [quantity, discount, profit, shipping_cost, year, order_date_year, order_date_month, order_date_week]
Index: []

In [97]:
df3.dropna(thresh=5) #Require that many non NAN values to not delete that row

Empty DataFrame
Columns: [quantity, discount, profit, shipping_cost, year, order_date_year, order_date_month, order_date_week]
Index: []

In [98]:
df3

quantity  discount  profit  shipping_cost  year  order_date_year  \
0           NaN       NaN  106.14            NaN  2011             2011   
2           NaN       NaN     NaN            NaN  2011             2011   
3           NaN       NaN     NaN            NaN  2011             2011   
4           NaN       NaN     NaN            NaN  2011             2011   
5           NaN       NaN     NaN            NaN  2011             2011   
...         ...       ...     ...            ...   ...              ...   
51285       NaN       NaN     NaN            NaN  2014             2014   
51286       NaN       NaN     NaN            NaN  2014             2014   
51287       NaN       NaN     NaN            NaN  2014             2014   
51288       NaN       NaN     NaN            NaN  2014             2014   
51289       NaN       NaN     NaN            NaN  2014             2014   

       order_date_month  order_date_week  
0                   NaN              NaN  
2                   NaN              NaN  
3                   NaN              NaN  
4                   NaN              NaN  
5                   NaN              NaN  
...                 ...              ...  
51285               NaN              NaN  
51286               NaN              NaN  
51287               NaN              NaN  
51288               NaN              NaN  
51289               NaN              NaN  

[51289 rows x 8 columns]

In [99]:
#Filling NAN values with 4
df3.fillna(value=4)

quantity  discount  profit  shipping_cost  year  order_date_year  \
0           4.0       4.0  106.14            4.0  2011             2011   
2           4.0       4.0    4.00            4.0  2011             2011   
3           4.0       4.0    4.00            4.0  2011             2011   
4           4.0       4.0    4.00            4.0  2011             2011   
5           4.0       4.0    4.00            4.0  2011             2011   
...         ...       ...     ...            ...   ...              ...   
51285       4.0       4.0    4.00            4.0  2014             2014   
51286       4.0       4.0    4.00            4.0  2014             2014   
51287       4.0       4.0    4.00            4.0  2014             2014   
51288       4.0       4.0    4.00            4.0  2014             2014   
51289       4.0       4.0    4.00            4.0  2014             2014   

       order_date_month  order_date_week  
0                   4.0              4.0  
2                   4.0              4.0  
3                   4.0              4.0  
4                   4.0              4.0  
5                   4.0              4.0  
...                 ...              ...  
51285               4.0              4.0  
51286               4.0              4.0  
51287               4.0              4.0  
51288               4.0              4.0  
51289               4.0              4.0  

[51289 rows x 8 columns]

In [102]:
df3.fillna(value = df3['profit']).mean()

quantity                    NaN
discount                    NaN
profit               285.710477
shipping_cost        205.753790
year                2012.777243
order_date_year     2012.777243
order_date_month            NaN
order_date_week             NaN
dtype: float64

In [104]:
(df3['profit']).mean()

285.71047695983384

In [105]:
#Average Profit year wise
df.groupby('order_date_year')['profit'].mean()

order_date_year
2011    27.665308
2012    28.043722
2013    29.604519
2014    28.758540
Name: profit, dtype: float64

In [107]:
#Average Shipping Cost year wise
df.groupby('order_date_year')['shipping_cost'].mean()

order_date_year
2011    27.149201
2012    25.861329
2013    26.418577
2014    26.268197
Name: shipping_cost, dtype: float64

In [106]:
#Average Discount year wise
df.groupby('order_date_year')['discount'].mean()

order_date_year
2011    0.148193
2012    0.141286
2013    0.140265
2014    0.143291
Name: discount, dtype: float64

In [115]:
#Average Sales year wise
df['sales'] = df.sales.str.replace(',',"").astype(int) #Converting into integer
df.groupby('order_date_year')['sales'].mean() 

order_date_year
2011    251.127153
2012    244.252235
2013    246.819335
2014    245.282129
Name: sales, dtype: float64

In [116]:
#Average Quantity year wise
df.groupby('order_date_year')['quantity'].mean() 

order_date_year
2011    3.494498
2012    3.476647
2013    3.488369
2014    3.457989
Name: quantity, dtype: float64

In [117]:
#Total Quantity year wise
df.groupby('order_date_year')['quantity'].sum() 

order_date_year
2011    31440
2012    38111
2013    48136
2014    60622
Name: quantity, dtype: int64

In [118]:
#Total Sales year wise
df.groupby('order_date_year')['sales'].sum()  #Sales and Quantity both are increasing. Good sign for business

order_date_year
2011    2259391
2012    2677493
2013    3405860
2014    4300041
Name: sales, dtype: int64

In [108]:
df.columns

Index(['order_id', 'order_date', 'ship_date', 'ship_mode', 'customer_name',
       'segment', 'state', 'country', 'market', 'region', 'product_id',
       'category', 'sub_category', 'product_name', 'sales', 'quantity',
       'discount', 'profit', 'shipping_cost', 'order_priority', 'year',
       'category1', 'converted_order_date', 'order_date_year',
       'order_date_month', 'order_date_week'],
      dtype='object')

In [111]:
df[['country','market']]

country  market
0            Algeria  Africa
2            Hungary    EMEA
3             Sweden      EU
4          Australia    APAC
5          Australia    APAC
...              ...     ...
51285  United States      US
51286        Morocco  Africa
51287      Nicaragua   LATAM
51288         Mexico   LATAM
51289  United States      US

[51289 rows x 2 columns]

**#CHALLENGES GIVEN**

Dataset 1: Agent Performance Data from iNeuron (Real Dataset)

Dataset 2: Agent Login Report


https://docs.google.com/spreadsheets/d/1z4EWsGxw0YBQ1-51xZDxYNlB3ngJzg4t/edit?usp=sharing&ouid=118282207943964605599&rtpof=true&sd=true


https://docs.google.com/spreadsheets/d/1z2n8aeFF5GoohQ3KCZw8bKogezpO9l4k/edit?usp=sharing&ouid=118282207943964605599&rtpof=true&sd=true

**TASKS:**

1. For each and every agent, Find out the weekly average rating for each agent (They workfor 5 days a week and not 7 days. So be careful with that. System will generate 7 records for a week)
2. Total Working Days for each Agent from Agent Performance Report
3. Total Queries taken by Agents
4. Total Feedback received by Agents
5. Agent Name who have average rating between 3.5 to 4
6. Agent Name who have rating less than 3.5
7. Agent Name who have rating more than 4.5
8. How many feedback agents have received more than Average
9. Average Weekly Response Time for each agent
10. Average Weekly Resolution Time for each agent
11. List of All Agents Name
12. Percentage of Chat on which they have recieved a feedback
13. Total Contribution hour for each and every agents weekly basis from Login Report Data
14. Total Percentage of Active Hour for a month   